addon/poll-task.js

```diff
@@ -0,0 +1,168 @@
+import Ember from 'ember';
+import { run } from '@ember/runloop';
+import { assert } from '@ember/debug';
+import getOrAllocate from './utils/get-or-allocate';
+import getTask from './utils/get-task';
+
+const { WeakMap } = Ember;
+
+/**
+ * A map of instances/poller functions that allows us to
+ * store poller tokens per instance.
+ *
+ * @private
+ *
+ */
+const registeredPollers = new WeakMap();
+
+let token = 0;
+let _shouldPollOverride;
+function shouldPoll() {
+  if (_shouldPollOverride) {
+    return _shouldPollOverride();
+  }
+
+  // eslint-disable-next-line ember-suave/no-direct-property-access
+  return !Ember.testing;
+}
+
+export function setShouldPoll(callback) {
+  _shouldPollOverride = callback;
+}
+
+let queuedPollTasks = Object.create(null);
+export function pollTaskFor(token) {
+  assert(
+    `You cannot advance pollTask '${token}' when \`next\` has not been called.`,
+    !!queuedPollTasks[token]
+  );
+
+  return run.join(null, queuedPollTasks[token]);
+}
+
+/**
+   Sets up a function that can perform polling logic in a testing safe way.
+   The task is invoked synchronously with an argument (generally called `next`).
+   In normal development/production when `next` is invoked, it will trigger the
+   task again (recursively). However, when in test mode the recursive polling
+   functionality is disabled, and usage of the `pollTaskFor` helper is required.
+
+   Example:
+
+   ```js
+   // app/components/foo-bar.js
+   export default Component.extend({
+     api: injectService(),
+
+     init() {
+       this._super(...arguments);
+
+       let token = this.pollTask((next) => {
+         this.get('api').request('get', 'some/path')
+           .then(() => {
+             this.runTask(next, 1800);
+           })
+       });
+
+       this._pollToken = token;
+     }
+   });
+   ```
+
+   Test Example:
+
+   ```js
+   import wait from 'ember-test-helpers/wait';
+   import { pollTaskFor } from 'ember-lifeline/mixins/run';
+
+   //...snip...
+
+   test('foo-bar watches things', function(assert) {
+     this.render(hbs`{{foo-bar}}`);
+
+     return wait()
+       .then(() => {
+         assert.equal(serverRequests, 1, 'called initially');
+
+         pollTaskFor(this._pollToken);
+         return wait();
+       })
+       .then(() => {
+         assert.equal(serverRequests, 2, 'called again');
+       });
+   });
+   ```
+
+   @method pollTask
+   @param { Function | String } taskOrName a function representing the task, or string
+                                           specifying a property representing the task,
+                                           which is run at the provided time specified
+                                           by timeout
+   @public
+   */
+export function pollTask(obj, taskOrName, token = getNextToken()) {
+  let next;
+  let task = getTask(obj, taskOrName, 'pollTask');
+  let tick = () => task.call(obj, next);
+
+  getOrAllocate(registeredPollers, obj, Array, getPollersDisposable).push(
```

Comment:
same general feels RE getOrAllocate here...

Reply:
Doesn't a bit of DRY help those feels?

Reply:
NO! 🙃 

IMHO, each abstraction should add value or it shouldn't exist. This specific one makes the actual logic _harder_ to understand...

Reply:
Fixed.